### PR TITLE
Pract 79/update get all decks controller public decks

### PR DESCRIPTION
**Pull request by koral14**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._

## Description

--updated getAllDecks controller to return only ‘public’ decks

--limited the flashcards per deck to max 50 in the Deck model

--added pagination to getDeckWithFlashcards Flashcards controller

--added a route for the getDeckWithFlashcards controller

--fixed the createFlashcard controller to findAndUpdate the deck when creating a new flashcard

*-- removed getDeck and getFlashcardsForDeck controllers and moved getDeckWithFlashcards to the Deck controllers

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
| ✓   | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before

image

### After

image

## Testing Steps / QA Criteria

Tested with Posman. Also debugged through different methods why created flashcards were not added to the deck.

© 2022 GitHub, Inc.
Terms
Privacy
Security
Status

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -16,11 +16,14 @@ const decksRouter = require('./routes/Decks.js');
 const allUnauthDecksRouter = require('./routes/decksAllUnauth');
 
 // middleware
-app.use(cors());
+app.use(cors({
+    origin: 'http://localhost:3000',
```
Comment (ddigioia3) [inline src/app.js:20]: [Configurability / Hardcoded Values]
This should be a configuration value in the .env.
That way if the front-end proxy address is something other than localhost:3000 the app still works.

File: src/controllers/Flashcards.js [hunk 2]
```diff
@@ -70,11 +78,24 @@ const getDeckWithFlashcards = async (req, res) => {
...
 };
 
-// (2) get the flashcards that appertain to a specific deck 
+// (2) get the flashcards that appertain to a specific deck - not used for the moment (with no detailed information about each flashcard)
 const getFlashcardsForDeck = async (req, res) => {
-    req.body.createdBy = req.user.userId;
-    const flashcards = await Flashcard.find({ createdBy: req.user.userId, deck: req.id }).sort('createdAt');
-    res.status(StatusCodes.CREATED).json({ flashcards, count: flashcards.length });
+    try {
+        req.body.createdBy = req.user.userId;
```
Comment (ddigioia3) [inline src/controllers/Flashcards.js:84]: [Requests Should Generally Be Immutable]
This doesn't seem to do anything? Can we remove this line? What was the intention?

File: src/app.js (hunk 1 above)
Comment (koral14) [reply src/app.js]: Added the path to the .env file. Thank you

File: src/routes/Flashcards.js [hunk 3]
```diff
@@ -6,10 +6,11 @@ const {
...
     createFlashcard,
     updateFlashcard,
-    deleteFlashcard
+    deleteFlashcard,
+    getDeckWithFlashcards,
 } = require('../controllers/Flashcards');
 
 router.route('/').post(createFlashcard).get(getUserFlashcards);
-router.route('/:id').get(getFlashcard).delete(deleteFlashcard).patch(updateFlashcard);
+router.route('/:id').get([getFlashcard, getDeckWithFlashcards]).delete(deleteFlashcard).patch(updateFlashcard);
```
Comment (ddigioia3) [inline src/routes/Flashcards.js:14]: [Functional Error]
If you define the functions in this way, it calls getFlashcard then getDeckWithFlashcards or calls it if getFlashcard calls next(). This isn't the intended behavior.

The user of the API needs to specify whether the id is for a deck or a flashcard.

As it stands, you may instead want to move this to the deck controller and create a separate route on the deck routes off `/deck/:id/flashcards`

**Review by ddigioia3 — CHANGES_REQUESTED** [review]

File: src/controllers/Flashcards.js (hunk 2 above)
Comment (koral14) [reply src/controllers/Flashcards.js:84]: There is no route for this controller either, as the previous controller does a similar job and I had an issue adding a separate route for this controller. So, I will remove the entire controller from here.

File: src/routes/Flashcards.js (hunk 3 above)
Comment (koral14) [reply src/routes/Flashcards.js:14]: I moved this route to the Decks routes, also moved its controller to the Decks controllers. I will test it now...

**Review by ddigioia3 — APPROVED** [review]

**Pull request merged**